frostsnap_core/src/device_nonces.rs

```diff
@@ -0,0 +1,379 @@
+}
+
+#[derive(Clone, Debug, PartialEq, bincode::Encode, bincode::Decode)]
+pub struct SigningState {
+    pub session_id: SignSessionId,
+    pub signature_shares: Vec<SignatureShare>,
+}
+
+pub trait NonceStreamSlot {
+    fn read_index(&mut self) -> Option<u32>;
+    fn read_slot(&mut self) -> Option<SecretNonceSlot>;
+    fn write_slot(&mut self, value: &SecretNonceSlot);
+}
+
+/// Manages two writable sectors of persistent storage such that we make sure the state of the system we're managing is never lost.
+/// The new state is first written, if that succeeds we finally write over the previous state.
+/// In this case we're managing the secret seed needed to produce nonces for signatures.
+#[derive(Clone, Debug, PartialEq)]
+pub struct ABSlot<S> {
+    slots: [S; 2],
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct ABSlots<S> {
+    slots: Vec<ABSlot<S>>,
+}
+
+impl<S: NonceStreamSlot> ABSlots<S> {
+    pub fn new(ab_slots: impl IntoIterator<Item = ABSlot<S>>) -> Self {
+        Self {
+            slots: ab_slots.into_iter().collect(),
+        }
+    }
+
+    pub fn get_or_create(
+        &mut self,
+        stream_id: NonceStreamId,
+        rng: &mut impl RngCore,
+    ) -> &mut ABSlot<S> {
+        let mut i = 0;
+        let found = loop {
+            let ab_slot = &mut self.slots[i];
+            match ab_slot.nonce_stream_id() {
+                Some(slot_nsid) => {
+                    if slot_nsid == stream_id {
+                        break Some(i);
+                    }
+                }
+                None => {
+                    ab_slot.initialize(stream_id, rng);
+                    break Some(i);
+                }
+            }
+            i += 1;
+            if i >= self.slots.len() {
+                break None;
+            }
+        };
+
+        match found {
+            Some(i) => &mut self.slots[i],
+            None => {
+                let random_overwrite_idx = rng.next_u32() as usize % self.slots.len();
+                let ab_slot = &mut self.slots[random_overwrite_idx];
+                ab_slot.initialize(stream_id, rng);
+                ab_slot
+            }
+        }
+    }
+
+    pub fn get(&mut self, stream_id: NonceStreamId) -> Option<&mut ABSlot<S>> {
+        //XXX: clippy is wrong about this
+        #[allow(clippy::manual_find)]
+        for slot in &mut self.slots {
+            if slot.nonce_stream_id() == Some(stream_id) {
+                return Some(slot);
+            }
+        }
+        None
+    }
+
+    pub fn all_stream_ids(&mut self) -> impl Iterator<Item = NonceStreamId> + '_ {
+        self.slots
+            .iter_mut()
+            .filter_map(|slot| slot.nonce_stream_id())
+    }
+}
+
+impl<S: NonceStreamSlot> ABSlot<S> {
+    pub fn new(a: S, b: S) -> Self {
+        Self { slots: [a, b] }
+    }
+
+    pub fn initialize(&mut self, stream_id: NonceStreamId, rng: &mut impl RngCore) {
+        let mut ratchet_prg_seed = ChaChaSeed::default();
```

Comment:
An occurrence of automatic trait impl derivation that is perhaps risky, think this works

```
#[derive(Clone, Debug, PartialEq, bincode::Encode, bincode::Decode)]
#[repr(transparent)]
pub struct ChaChaSeed([u8; 32]);

impl ChaChaSeed {
    pub fn new(rng: &mut impl RngCore) -> Self {
        let mut bytes = [0; 32];
        rng.fill_bytes(&mut bytes);
        Self(bytes)
    }  
}
```